Stateful2/MyCommunicationListener.cs

```diff
@@ -43,4 +53,93 @@ public void Abort()
+            }
+            return random.Next(0, 300);
+        }
+    }
+
+    class ServerSideDistributionFeature : Feature
+    {
+        public ServerSideDistributionFeature()
+        {
+            EnableByDefault();
+        }
+
+        protected override void Setup(FeatureConfigurationContext context)
+        {
+            context.Pipeline.Register(b => new ServerSideDistributor(
+                b.Build<StatefulServiceContext>(), 
+                address => context.Settings.Get<TransportInfrastructure>().ToTransportAddress(address), 
+                context.Settings.LogicalAddress(),
+                context.Settings.Get<Guid>("PartitionId")), "Foo");
+        }
+    }
+
+    class ServerSideDistributor : IBehavior<IIncomingLogicalMessageContext, IIncomingLogicalMessageContext>
+    {
+        private StatefulServiceContext statefulServiceContext;
+        private Func<LogicalAddress, string> addressTranslator;
+        private Guid currentPartition;
+        private LogicalAddress logicalAddress;
+        private readonly ServicePartitionResolver servicePartitionResolver = ServicePartitionResolver.GetDefault();
+
+        public ServerSideDistributor(StatefulServiceContext context, Func<LogicalAddress, string> addressTranslator, LogicalAddress local, Guid currentPartition)
+        {
+            logicalAddress = local;
+            this.currentPartition = currentPartition;
+            this.addressTranslator = addressTranslator;
+            statefulServiceContext = context;
+        }
+
+        public async Task Invoke(IIncomingLogicalMessageContext context, Func<IIncomingLogicalMessageContext, Task> next)
+        {
+            var contextMessage = context.Message;
+
+            if (contextMessage.MessageType != typeof(MyNotProperlyRoutedMessage))
+            {
+                await next(context).ConfigureAwait(false);
+                return;
+            }
+
+            var partitionValue = PartitionMap.Map(contextMessage.Instance);
+            // Think about custom header or smart caching...
+            var partitionKey = partitionValue is string
+                ? new ServicePartitionKey((string) partitionValue)
+                : new ServicePartitionKey((int) partitionValue);
+            var partition1 =
+                await this.servicePartitionResolver.ResolveAsync(statefulServiceContext.ServiceName, partitionKey,
+                    CancellationToken.None).ConfigureAwait(false);
+
+
+            if (partition1.Info.Id != currentPartition)
+            {
+                ServiceEventSource.Current.ServiceMessage(statefulServiceContext,
+                    $"Received MyNotProperlyRoutedMessage on partition {currentPartition} rerouting to {partition1.Info.Id}");
+                var destination =
+                    addressTranslator(logicalAddress.CreateIndividualizedAddress(partition1.Info.Id.ToString()));
+                await context.ForwardCurrentMessageTo(destination).ConfigureAwait(false);
```

Comment:
Discussed this with Bob. It will result in unnecessary operations (read cost). Not sure how much of a concern it truly is though.